sql/analyzer/resolve_ctes.go

```diff
@@ -216,8 +216,8 @@ func hoistCommonTableExpressions(ctx *sql.Context, a *Analyzer, n sql.Node, scop
 			if err != nil {
 				return nil, transform.SameTree, err
 			}
-			if _, isCTE := l.(*plan.With); isCTE {
-				return hoistCommonTableExpressions(ctx, a, plan.NewUnion(l, r), scope, sel)
+			if cte, isCTE := l.(*plan.With); isCTE {
```

Comment:
I'm not sure this is right. On line 208, we explicitly don't return `newWith` if `cte.Recursive`, and this edit changes default CTE behavior. Seems there is a reason we continue to hoist the children of ctes, why for default CTEs recurse in the way they do, and that we should instead switch all returns on whether we know we are in a recursive CTE or not. @reltuk could you check me on this?

Reply:
the uptick in recursive CTE bugs also suggests that we need to test these more intensively, like 30-50 more tests, with nesting, group by, joins, order by, limit, having, window, subqueries, union, etc

Reply:
While I agree with the sentiment, i would not gate this fix on adding those tests. We could make a new issue to add those tests.

Reply:
Obviously, add as many tests as you need to make sure this change is safe. Not sure if that is 30-50.

Reply:
Yeah I agree, I mainly said that because I wish I could pinpoint how this is suspect in a more useful way.

Reply:
Here's some examples that my be good to codify as tests. https://learnsql.com/blog/cte-with-examples/

Reply:
Looks like the tests cover these cases and more.

Reply:
Just for this PR, a change that preserves preexisting CTE behavior and fixes just the recursive panic is safe enough to merge without a lot more testing. But generally we need to test each terminal line of the function, the recursive tails and base cases. So unions with CTEs that themselves have unions that need hoisting. With top-level and base cases for both CTEs and recursive CTEs.